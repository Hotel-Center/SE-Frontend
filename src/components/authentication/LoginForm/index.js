import { Box, Button, Grid, TextField } from "@mui/material";
import { useFormik } from "formik";
import Link from "next/link";
import React from "react";
import * as yup from "yup";

export default function LoginForm(props) {
  const formik = useFormik({
    initialValues: {
      email: "",
      password: "",
    },
    onSubmit: props.handleSubmit,
    validationSchema: yup.object({
      email: yup
        .string()
        .required("Please enter your email address")
        .email("Please enter a valid email address"),
      password: yup.string().required("Please enter your password"),
    }),
  });

  return (
    <Box
      role="form"
      component="form"
      noValidate
      onSubmit={formik.handleSubmit}
      sx={{ mt: 1 }}
    >
      <Grid container spacing={1}>
        <Grid item xs={12}>
          <TextField
            required
            fullWidth
            id="email"
            label="Email Address"
            name="email"
            autoComplete="email"
            value={formik.values.email}
            onChange={formik.handleChange}
            onBlur={formik.handleBlur}
            error={
              Boolean(formik.touched.email) && Boolean(formik.errors.email)
            }
            helperText={(formik.touched.email && formik.errors.email) || " "}
          />
        </Grid>

        <Grid item xs={12}>
          <TextField
            required
            fullWidth
            name="password"
            label="Password"
            role="password"
            id="password"
            autoComplete="password"
            value={formik.values.password}
            onBlur={formik.handleBlur}
            onChange={formik.handleChange}
            error={
              Boolean(formik.touched.password) &&
              Boolean(formik.errors.password)
            }
            helperText={
              (formik.touched.password && formik.errors.password) || " "
            }
          />
        </Grid>

        {/* <div>
        <h5 styles={{ color: "red", fontSize: "14.5px" }}>
          {formik.errors.email}
        </h5>
      </div> */}
      </Grid>
      <Button
        type="submit"
        fullWidth
        variant="contained"
        // onClick={formik.handleSubmit}
        sx={{
          mt: 2,
          mb: 2,
          marginTop: "20px",
          height: "43px",
          marginBottom: "10px",
        }}
      >
        Login
      </Button>

      <Grid container justifyContent="flex-end">
        <Grid item>
          <Link href="/sign-up" className="lnk-login">
            {"Don't have an account yet? Sign Up"}
          </Link>
        </Grid>
      </Grid>
    </Box>
  );
}
