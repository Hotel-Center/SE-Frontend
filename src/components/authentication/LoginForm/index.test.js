import "@testing-library/jest-dom/extend-expect";
import { render, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import LoginForm from ".";

describe("Login Form", () => {
  it("renders Login button", () => {
    const loginForm = render(<LoginForm />);

    expect(loginForm.getByRole("button", { text: "Login" })).toBeTruthy();
  });

  it("renders login fields", () => {
    const loginForm = render(<LoginForm />);

    expect(
      loginForm.getByRole("textbox", { name: "Email Address" })
    ).toBeTruthy();
    expect(loginForm.getByRole("password", { name: "Password" })).toBeTruthy();
  });

  it("has editable login fields", async () => {
    const user = userEvent.setup();
    const loginForm = render(<LoginForm />);

    let emailInput = loginForm.getByRole("textbox", {
      name: "Email Address",
    });
    let passwordField = loginForm.getByRole("password", {
      name: "Password",
    });
    let passwordInput = passwordField.children[1].firstElementChild;

    expect(emailInput.textContent).toBe("");
    expect(passwordInput.textContent).toBe("");

    const email = "email@example.com";
    const password = "example_password";

    await user.type(emailInput, email);
    await user.type(passwordInput, password);

    expect(emailInput).toHaveValue(email);
    expect(passwordInput).toHaveValue(password);
  });

  it("ignores the email validation if email input is not onfocused", async () => {
    const user = userEvent.setup();
    const loginForm = render(<LoginForm />);
    let emailInput = loginForm.getByRole("textbox", {
      name: "Email Address",
    });

    await user.type(emailInput, "invalid@email");

    expect(loginForm.queryByText("Please enter your email address")).toBeNull();
  });

  it("ignores the password validation if password input is not onfocused", async () => {
    const user = userEvent.setup();
    const loginForm = render(<LoginForm />);
    let passwordField = loginForm.getByRole("password", {
      name: "Password",
    });

    await user.click(passwordField);

    expect(loginForm.queryByText("Please enter your password")).toBeNull();
  });

  it("fails the email validation if email is empty", async () => {
    const user = userEvent.setup();
    const loginForm = render(<LoginForm />);
    let emailInput = loginForm.getByRole("textbox", {
      name: "Email Address",
    });

    await user.click(emailInput);
    await waitFor(() => {
      emailInput.blur();
      expect(emailInput).not.toHaveFocus();
    });

    expect(
      await loginForm.findByText("Please enter your email address")
    ).toBeTruthy();
  });

  it("fails the email validation if email is invalid", async () => {
    const user = userEvent.setup();
    const loginForm = render(<LoginForm />);
    let emailInput = await loginForm.findByRole("textbox", {
      name: "Email Address",
    });

    await user.type(emailInput, "invalid@email");
    await waitFor(() => {
      emailInput.blur();
      expect(emailInput).not.toHaveFocus();
    });

    expect(
      loginForm.queryByText("Please enter a valid email address")
    ).toBeTruthy();
  });

  it("fails the password validation if password is empty", async () => {
    const user = userEvent.setup();
    const loginForm = render(<LoginForm />);
    let passwordField = loginForm.getByRole("password", {
      name: "Password",
    });
    let passwordInput = passwordField.children[1].firstElementChild;

    await user.click(passwordInput);
    await waitFor(() => {
      passwordInput.blur();
      expect(passwordInput).not.toHaveFocus();
    });

    expect(loginForm.queryByText("Please enter your password")).toBeTruthy();
  });

  it("ignores password validation if email is validated", async () => {
    const user = userEvent.setup();
    const loginForm = render(<LoginForm />);
    let emailInput = loginForm.getByRole("textbox", {
      name: "Email Address",
    });

    await user.click(emailInput);
    await waitFor(() => {
      emailInput.blur();
      expect(emailInput).not.toHaveFocus();
    });

    expect(loginForm.queryByText("Please enter your password")).toBeNull();
  });

  it("submits valid inputs", async () => {
    const handleSubmit = jest.fn();
    const user = userEvent.setup();
    const loginForm = render(<LoginForm handleSubmit={handleSubmit} />);
    let emailInput = loginForm.getByRole("textbox", {
      name: "Email Address",
    });
    let passwordField = loginForm.getByRole("password", {
      name: "Password",
    });
    let submitButton = loginForm.getByRole("button", {
      name: "Login",
    });
    let passwordInput = passwordField.children[1].firstElementChild;

    await user.type(emailInput, "valid@email.address");
    await user.type(passwordInput, "password");
    await waitFor(() => {
      passwordInput.blur();
      expect(passwordInput).not.toHaveFocus();
    });

    expect(loginForm.queryByText("Please enter your email address")).toBeNull();
    expect(
      loginForm.queryByText("Please enter a valid email address")
    ).toBeNull();
    expect(loginForm.queryByText("Please enter your password")).toBeNull();

    await user.click(submitButton);

    expect(handleSubmit).toHaveBeenCalledTimes(1);
    expect(handleSubmit).toHaveBeenCalledWith(
      {
        email: "valid@email.address",
        password: "password",
      },
      expect.anything()
    );
  });
});
