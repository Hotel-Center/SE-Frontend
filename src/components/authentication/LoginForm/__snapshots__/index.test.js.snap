// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Form renders correctly 1`] = `
<form
  className="MuiBox-root css-164r41r"
  noValidate={true}
  onSubmit={[Function]}
  role="form"
>
  <div
    className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-1 css-a5rdam-MuiGrid-root"
  >
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1idn90j-MuiGrid-root"
    >
      <div
        className="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
      >
        <label
          className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary Mui-required MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink={false}
          htmlFor="email"
          id="email-label"
        >
          Email Address
          <span
            aria-hidden={true}
            className="MuiFormLabel-asterisk MuiInputLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
          >
             
            *
          </span>
        </label>
        <div
          className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
          onClick={[Function]}
        >
          <input
            aria-describedby="email-helper-text"
            aria-invalid={false}
            autoComplete="email"
            autoFocus={false}
            className="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
            disabled={false}
            id="email"
            name="email"
            onAnimationStart={[Function]}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            required={true}
            type="text"
            value=""
          />
          <fieldset
            aria-hidden={true}
            className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              className="css-yjsfm1"
            >
              <span>
                Email Address
                 
                *
              </span>
            </legend>
          </fieldset>
        </div>
        <p
          className="MuiFormHelperText-root MuiFormHelperText-sizeMedium MuiFormHelperText-contained Mui-required css-1wc848c-MuiFormHelperText-root"
          id="email-helper-text"
        >
          <span
            className="notranslate"
          >
            ​
          </span>
        </p>
      </div>
    </div>
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1idn90j-MuiGrid-root"
    >
      <div
        aria-label="Password"
        className="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
        role="password"
      >
        <label
          className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary Mui-required MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink={false}
          htmlFor="password"
          id="password-label"
        >
          Password
          <span
            aria-hidden={true}
            className="MuiFormLabel-asterisk MuiInputLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
          >
             
            *
          </span>
        </label>
        <div
          className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
          onClick={[Function]}
        >
          <input
            aria-describedby="password-helper-text"
            aria-invalid={false}
            autoComplete="password"
            autoFocus={false}
            className="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
            disabled={false}
            id="password"
            name="password"
            onAnimationStart={[Function]}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            required={true}
            type="password"
            value=""
          />
          <fieldset
            aria-hidden={true}
            className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              className="css-yjsfm1"
            >
              <span>
                Password
                 
                *
              </span>
            </legend>
          </fieldset>
        </div>
        <p
          className="MuiFormHelperText-root MuiFormHelperText-sizeMedium MuiFormHelperText-contained Mui-required css-1wc848c-MuiFormHelperText-root"
          id="password-helper-text"
        >
          <span
            className="notranslate"
          >
            ​
          </span>
        </p>
      </div>
    </div>
  </div>
  <button
    className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-fullWidth MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-fullWidth css-111ww5k-MuiButtonBase-root-MuiButton-root"
    disabled={false}
    onBlur={[Function]}
    onContextMenu={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={0}
    type="submit"
  >
    Login
  </button>
  <div
    className="MuiGrid-root MuiGrid-container css-16g5rpk-MuiGrid-root"
  >
    <div
      className="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
    >
      <a
        className="lnk-login"
        href="/sign-up"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onTouchStart={[Function]}
      >
        Don't have an account yet? Sign Up
      </a>
    </div>
  </div>
</form>
`;
